artifact/signer_test.go

```diff
@@ -0,0 +1,241 @@
+XbeslVVE29ZP7iJNkVT1JxFWkfA3/gQXn8h0or87wPGu+bX4jw6BK46mP717RgCT
+0dlFBsy2xqtcPzkiW6Sx4pqjYUQC37TJ63/vvXkPlvFUpUzmGzZ9V5mQLupwtQ2z
+MIXnMqXyHgMqtVA=
+-----END PUBLIC KEY-----`
+	PrivateDSAKey = `-----BEGIN DSA PRIVATE KEY-----
+MIIBuwIBAAKBgQCjXKmtUP9LN1vDlt2JVYA9krxVAGSb9zeb58MVJcSDOFT0feHD
+cTkBPU60jCNNNbcHAQuMmDNVgRjXEC4cLgOxAM55I46ULi3m8lBCmQXdnCTJD6QR
+uWBfljiWlzF9SU7bjEC4pP7yZX20Iyuw20+uNBb3tgfvpEYqupatppqTiQIVANV5
+9bgDNAR+rTtn1CjNnuNyI377AoGBAJQZscbF0hsLVCbYqZRoM8hocpqi1zpE3/gw
+vAHMw09IScRNIj1U+2aCb4LMikji0iURJ3XoqUkjl8kUraET1JZNkUmGAr10xs8H
+8XUtVqMmBb64o/8BZWCG700CaAd+sEwyO/06fv3/YonvuR1nBS+jRpnkvjfH89DI
+IsFZHZFqAoGANzjN9AfCzxcAswYvZyDn3bHR9FoaXbeslVVE29ZP7iJNkVT1JxFW
+kfA3/gQXn8h0or87wPGu+bX4jw6BK46mP717RgCT0dlFBsy2xqtcPzkiW6Sx4pqj
+YUQC37TJ63/vvXkPlvFUpUzmGzZ9V5mQLupwtQ2zMIXnMqXyHgMqtVACFFTsC83p
+BlUY/oCrAGUGN10F49+c
+-----END DSA PRIVATE KEY-----`
+)
+
+func TestPublicKey(t *testing.T) {
+	m, err := getKeyAndVerifyMethod([]byte(PublicRSAKey))
+	assert.NoError(t, err)
+	assert.NotNil(t, m)
+	assert.IsType(t, &RSA{}, m.method)
+	assert.IsType(t, &rsa.PublicKey{}, m.key)
+
+	m, err = getKeyAndVerifyMethod([]byte(PublicECDSAKey))
+	assert.NoError(t, err)
+	assert.NotNil(t, m)
+	assert.IsType(t, &ECDSA256{}, m.method)
+	assert.IsType(t, &ecdsa.PublicKey{}, m.key)
+
+	m, err = getKeyAndVerifyMethod([]byte(PublicDSAKey))
+	assert.Error(t, err)
+	assert.Nil(t, m)
+
+	m, err = getKeyAndVerifyMethod([]byte("some ivalid key"))
+	assert.Error(t, err)
+	assert.Nil(t, m)
+	m, err = getKeyAndVerifyMethod([]byte(PublicRSAKeyInvalid))
+	assert.Error(t, err)
+	assert.Nil(t, m)
+}
+
+func TestPrivateKey(t *testing.T) {
+	m, err := getKeyAndSignMethod([]byte(PrivateRSAKey))
+	assert.NoError(t, err)
+	assert.NotNil(t, m)
+	assert.IsType(t, &RSA{}, m.method)
+	assert.IsType(t, &rsa.PrivateKey{}, m.key)
+
+	m, err = getKeyAndSignMethod([]byte(PrivateECDSAKey))
+	assert.NoError(t, err)
+	assert.NotNil(t, m)
+	assert.IsType(t, &ECDSA256{}, m.method)
+	assert.IsType(t, &ecdsa.PrivateKey{}, m.key)
+
+	m, err = getKeyAndSignMethod([]byte(PrivateDSAKey))
+	assert.Error(t, err)
+	assert.Nil(t, m)
+
+	m, err = getKeyAndSignMethod([]byte("invalid key"))
+	assert.Error(t, err)
+	assert.Nil(t, m)
+}
+
+func TestRSA(t *testing.T) {
+	msg := []byte("this is secret message")
+
+	s := NewSigner([]byte(PrivateRSAKey))
+	sig, err := s.Sign(msg)
+	assert.NoError(t, err)
+	assert.NotNil(t, sig)
+
+	v := NewVerifier([]byte(PublicRSAKey))
+	err = v.Verify(msg, sig)
+	assert.NoError(t, err)
+
+	// use invalid key
+	v = NewVerifier([]byte(PublicRSAKeyError))
+	err = v.Verify(msg, sig)
+	assert.Error(t, err)
+	assert.Contains(t, errors.Cause(err).Error(), "verification error")
+}
+
+func TestRSARaw(t *testing.T) {
+	r := RSA{}
+	sig, err := r.Sign([]byte("my message"), PublicRSAKey)
+	assert.Error(t, err)
+	assert.Nil(t, sig)
+
+	err = r.Verify([]byte("my message"), []byte("signature"), PrivateRSAKey)
+	assert.Error(t, err)
+}
+
+func TestECDSA(t *testing.T) {
```

Comment:
You should also test that an verification fails with an invalid signature. For an invalid signature, just modify one random byte of the valid signature. ;-)